File: SDK/src/com/loment/cashewnut/enc/Crypter.java
package com.loment.cashewnut.enc;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import org.bouncycastle.crypto.DataLengthException;
import org.bouncycastle.crypto.engines.AESFastEngine;
import org.bouncycastle.crypto.modes.CBCBlockCipher;
import org.bouncycastle.crypto.params.KeyParameter;
import org.bouncycastle.crypto.params.ParametersWithIV;

import com.loment.cashewnut.MyException;

/**
 * @author Madhur Utility class to encrypt decrypt data using AES. Usage: From
 *         any of the datastore classes use this code: byte[] encdata =
 *         Crypter.encrypt(new RecordKey(),data);
 * 
 *         For Message encryption use: String passphrase =
 *         target.substring(target.length()-4,4) +
 *         phoneNumber.substring(phoneNumber.length()-4,4); byte[] encdata =
 *         Crypter.encrypt(new MessageKey(passphrase),message.getBytes());
 * 
 */
public class Crypter {
	private static CBCBlockCipher encryptEngine = new CBCBlockCipher(
			new AESFastEngine());
	private static CBCBlockCipher decryptEngine = new CBCBlockCipher(
			new AESFastEngine());

	private static byte[] initVector = new byte[] { 0, 1, 2, 3, 4, 5, 6, 7, 8,
			9, 11, 12, 13, 14, 15, 16 };

	/**
	 * @param key
	 *            Any of the key generating implementation classes
	 * @param data
	 *            Data to decrypt
	 * @return decrypted data
	 */
	public synchronized static byte[] decrypt(Key key, byte[] data) {

		try {
			ParametersWithIV param = new ParametersWithIV(new KeyParameter(
					key.getKey()), initVector);
			byte[] clear = new byte[data.length];

			decryptEngine.init(false, param);
			int blocksize = decryptEngine.getBlockSize();

			for (int i = 0; i < data.length; i += blocksize) {
				decryptEngine.processBlock(data, i, clear, i);
			}
			decryptEngine.reset();
			ByteArrayInputStream inStream = new ByteArrayInputStream(clear);
			DataInputStream dataStream = new DataInputStream(inStream);

			try {
				int size = dataStream.readShort();
				// added extra check for data which is larger then max size of
				// short.
				// this was causing few messages with large content not to load.
				if (size < 0 || size > clear.length
						|| size < (clear.length - (blocksize + 1))) {
					size = clear.length - 2;
				}
				byte[] out = new byte[size];
				dataStream.read(out);
				dataStream.close();
				return out;
			} catch (IOException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			return new byte[] { (byte) 0xff };
		} catch (Exception e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
			return new byte[] { (byte) 0xff };
		}
	}

	/**
	 * @param key
	 *            Any of the key generating implementation classes
	 * @param data
	 *            Data to encrypt
	 * @return encrypted data
	 */
	public synchronized static byte[] encrypt(Key key, byte[] data) {
		ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
		DataOutputStream dataStream = new DataOutputStream(byteStream);
		ParametersWithIV param = new ParametersWithIV(new KeyParameter(
				key.getKey()), initVector);

		encryptEngine.init(true, param);
		int blocksize = encryptEngine.getBlockSize();
		try {
			dataStream.writeShort(data.length);
			dataStream.write(data);
			dataStream.close();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
			return null;
		}

		byte[] in = new byte[((byteStream.size() + blocksize - 1) / blocksize)
				* blocksize];
		System.arraycopy(byteStream.toByteArray(), 0, in, 0, byteStream.size());
		byte[] out = new byte[in.length];

		for (int i = 0; i < in.length; i += blocksize) {
			encryptEngine.processBlock(in, i, out, i);
		}
		encryptEngine.reset();
		return out;
	}

	/**
	 * @param key
	 *            Any of the key generating implementation classes
	 * @param data
	 *            Data to decrypt
	 * @return decrypted data
	 * @throws MyException 
	 */
	public synchronized static byte[] decryptNew(Key key, byte[] data, 
			boolean hasPadding) throws MyException {

		try {
			ParametersWithIV param = new ParametersWithIV(new KeyParameter(
					key.getKey()), initVector);
			byte[] clear = new byte[data.length];

			decryptEngine.init(false, param);
			int blocksize = decryptEngine.getBlockSize();

			for (int i = 0; i < data.length; i += blocksize) {
				decryptEngine.processBlock(data, i, clear, i);
			}
			decryptEngine.reset();
			ByteArrayInputStream inStream = new ByteArrayInputStream(clear);
			// DataInputStream dataStream = new DataInputStream(inStream);

			try {
				if (hasPadding) {
					int paddingSize = 0;
					for (int i = clear.length - 1; i >= clear.length
							- blocksize; i--) {
						if (clear[i] == 0)
							paddingSize++;
						else
							break;
					}

					byte[] out = new byte[clear.length - paddingSize];
					inStream.read(out);
					return out;
				} else {
					return clear;
				}
			} catch (IOException e) {
				try {
					throw e;
				} catch (Exception e1) {
					e1.printStackTrace();
				}
			}
		} catch (Exception e) {
			throw new MyException(e.getMessage());
		}
		return new byte[] { (byte) 0xff };
	}

	/**
	 * @param key
	 *            Any of the key generating implementation classes
	 * @param data
	 *            Data to encrypt
	 * @return encrypted data
	 */
	public synchronized static byte[] encryptNew(Key key, byte[] data) {
		ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
		ParametersWithIV param = new ParametersWithIV(new KeyParameter(
				key.getKey()), initVector);

		encryptEngine.init(true, param);
		int blocksize = encryptEngine.getBlockSize();
		try {
			byteStream.write(data);
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
			return null;
		}
		int blocks = data.length / blocksize;
		if ((data.length % blocksize) > 0) {
			blocks++;
		}
		byte[] in = new byte[blocksize * blocks];
		// byte[] in = new byte[((byteStream.size() + blocksize - 1) /
		// blocksize) * blocksize];
		System.arraycopy(byteStream.toByteArray(), 0, in, 0, byteStream.size());
		byte[] out = new byte[in.length];

		for (int i = 0; i < in.length; i += blocksize) {
			encryptEngine.processBlock(in, i, out, i);
		}
		encryptEngine.reset();
		return out;
	}
}

File: SDK/Cashew/Internal/Encryption/Crypter.cs
using Org.BouncyCastle.Crypto.Engines;
using Org.BouncyCastle.Crypto.Modes;
using Org.BouncyCastle.Crypto.Parameters;
using System;
using System.Diagnostics;
using System.IO;

namespace Cashew.Internal.Encryption
{
    internal class Crypter
    {
        /// <summary>
        /// Used for decrypting
        /// </summary>
        private static readonly byte[] InitVector = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 11, 12, 13, 14, 15, 16 };

        /// <summary>
        /// The engine used to decrypt SHA256 encoded data
        /// </summary>
        private static readonly CbcBlockCipher DecryptCipher = new CbcBlockCipher(new AesEngine());

        /// <summary>
        /// The engine used to encrypt SHA256 encoded data
        /// </summary>
        private static readonly CbcBlockCipher EncryptCipher = new CbcBlockCipher(new AesEngine());

        /// <param name="key">The key to be used to decrypt the data.</param>
        /// <param name="data">The encrypted data.</param>
        /// <param name="hasPadding">Whether the encrypted data includes padding.</param>
        /// <returns>The decrypted data.</returns>
        internal static byte[] DecryptNew(IKey key, byte[] data, bool hasPadding = true)
        {
            var param = new ParametersWithIV(new KeyParameter(key.GetKey()), InitVector);
            var clear = new byte[data.Length];

            DecryptCipher.Init(false, param);
            var blocksize = DecryptCipher.GetBlockSize();

            for (var i = 0; i < data.Length; i += blocksize)
            {
                DecryptCipher.ProcessBlock(data, i, clear, i);
            }

            DecryptCipher.Reset();

            var inStream = new MemoryStream(clear);

            if (hasPadding)
            {
                var paddingSize = 0;
                for (var i = clear.Length - 1; i >= clear.Length - blocksize; i--)
                {
                    if (clear[i] == 0)
                        paddingSize++;
                    else
                        break;
                }

                var outb = new byte[clear.Length - paddingSize];
                inStream.Read(outb, 0, outb.Length);
                return outb;
            }

            return clear;
        }

        /// <param name="key">The key to be used to encrypt the data.</param>
        /// <param name="data">The data to be encrypted.</param>
        /// <returns>The encrypted data.</returns>
        internal static byte[] EncryptNew(IKey key, byte[] data)
        {
            try
            {
                var byteStream = new MemoryStream();
                var keyParam = new KeyParameter(key.GetKey(), 0, 32);
                var param = new ParametersWithIV(keyParam, InitVector);

                EncryptCipher.Init(true, param);

                var blocksize = EncryptCipher.GetBlockSize();

                byteStream.Write(data, 0, data.Length);

                var blocks = data.Length / blocksize;

                if ((data.Length % blocksize) > 0)
                {
                    blocks++;
                }

                var inb = new byte[blocksize * blocks];

                Array.Copy(byteStream.ToArray(), 0, inb, 0, (int)(byteStream.Length));

                var outb = new byte[inb.Length];

                for (var i = 0; i < inb.Length; i += blocksize)
                {
                    EncryptCipher.ProcessBlock(inb, i, outb, i);
                }

                EncryptCipher.Reset();

                return outb;
            }
            catch (Exception e)
            {
                Debug.WriteLine(e.StackTrace);

                return new byte[] { 0xff };
            }
        }
    }
}
